### DISCUSSION TITLE
关于使用net.apply_feature_transform的归一化问题

### DISCUSSION THREAD

**Discussion by Little7776** [lululxvi/deepxde#791] [Q&A]
博士您好
我尝试对板子的受力弯曲微分方程进行分析,现在遇到了一些困难
1.板子长宽a=480 ，b=240
我在设定好微分方程后，不知道是不是因为量级太大了所以输出效果很不好，还出现点的跳动情况。
![QQ图片20220711112817](https://user-images.githubusercontent.com/73152456/178212747-99e8767d-7852-47e9-8ebf-752b6dd6e5d6.png)

2.之后我尝试用net.apply_feature_transform(lambda X: tf.concat([X[:,0:1]/a, X[:,1:2]/b], axis=1))对输
入进行归一化，得到波动图像，因此我认为归一化有比较好的效果，但是依旧不满足用matlab和有限元验证的结果，
![QQ图片20220711112951](https://user-images.githubusercontent.com/73152456/178212933-c332669f-fc8c-401c-aa7b-db0a580046e0.png)
matlab结果
![QQ图片20220711112832](https://user-images.githubusercontent.com/73152456/178212984-1338f033-5fa5-48db-a810-b6699dd083c5.png)
使用net.apply_feature_transform的结果
3.在此基础上，我对边界进行处理，将狄利克雷边界条件改为硬边界约束，得到较为相近的图像结果，但是会缺少一条边的结果，在硬边界的基础上我加上原来的狄利克雷边界条件则边界没有问题，请问这个做法可以吗？
使用net.apply_feature_transform对输入进行比例缩放后为什么输出图像的x并没有发生变化，而且产生图像的范围也没有改变？
当把net.apply_feature_transform中的a,b换成小于10的值，误差图像中会有训练损失曲线，但是大于10之后就只有测试损失曲线，这是为什么呢？
![QQ图片20220711112853](https://user-images.githubusercontent.com/73152456/178213220-a0a432a6-a100-41cd-a487-85d5682dc70a.png)
使用硬边界去掉狄利克雷边界条件
![QQ图片20220711112844](https://user-images.githubusercontent.com/73152456/178213323-e0125f33-5475-440d-800a-c3be35cedb64.png)
使用硬边界且加上狄利克雷边界条件
![QQ图片20220711154154](https://user-images.githubusercontent.com/73152456/178213505-7471b181-539c-4e7e-a104-8b147779f9ea.png)
net.apply_feature_transform中当数值大于10时

以下是部分代码，希望能得到您的回复
`def pde(x,y):

    dw_xx = dde.grad.hessian(y, x, i=0, j=0)
    dw_yy = dde.grad.hessian(y, x, i=1, j=1)
    dw_xxxx = dde.grad.hessian(dw_xx, x, i=0, j=0)
    dw_xxyy = dde.grad.hessian(dw_xx, x, i=1, j=1)
    dw_yyyy = dde.grad.hessian(dw_yy, x, i=1, j=1)
    return dw_xxxx + 2 * dw_xxyy + dw_yyyy - q / D

def transform(x, y):
    res = x[:, 0:1] * (a - x[:, 0:1]) * x[:, 1:2] * (b - x[:, 1:2])
    return res * y

def boundary(x, on_boundary):
    return on_boundary

def boundary_x(x,on_boundary):
    return on_boundary and np.isclose(x[0],0 or a)

def boundary_y(x,on_boundary):
    return on_boundary and np.isclose(x[1],0 or b)

geom=dde.geometry.geometry_2d.Rectangle(xmin=[0,0], xmax=[a,b])

bc_1=dde.icbc.DirichletBC(geom,lambda x:0,boundary)
bc_x=dde.icbc.OperatorBC(geom,lambda x,y,_:ddw(x,y)[0],boundary_x)
bc_y=dde.icbc.OperatorBC(geom,lambda x,y,_:ddw(x,y)[1],boundary_y)

data = dde.data.PDE(geom, pde, [bc_1,bc_x,bc_y], num_domain=1200, num_boundary=240, num_test=1500)
net = dde.nn.FNN([2] + [50] * 4 + [1], "tanh", "Glorot uniform")

net.apply_feature_transform(lambda X: tf.concat([X[:,0:1]/a, X[:,1:2]/b], axis=1))

net.apply_output_transform(transform)

model = dde.Model(data, net)

model.compile("adam", lr=0.001)
model.train(epochs=5000)
model.compile("L-BFGS")
`

**Comment by lululxvi** [comment]
> 3.在此基础上，我对边界进行处理，将狄利克雷边界条件改为硬边界约束，得到较为相近的图像结果，但是会缺少一条边的结果，在硬边界的基础上我加上原来的狄利克雷边界条件则边界没有问题，请问这个做法可以吗？ 

Yes.

> 使用net.apply_feature_transform对输入进行比例缩放后为什么输出图像的x并没有发生变化，而且产生图像的范围也没有改变？ 

`net.apply_feature_transform` doesn't change the input and output. It is a transformation inside the network.

> 当把net.apply_feature_transform中的a,b换成小于10的值，误差图像中会有训练损失曲线，但是大于10之后就只有测试损失曲线，这是为什么呢？

The two lines are overlapped.

**Reply by Little7776** [reply]
非常感谢您的回答
但是我还有两个疑问，希望您能指点一二
1.当把net.apply_feature_transform中的a,b换成小于10的值，误差图像中会有训练损失曲线，但是大于10之后就只有测试损失曲线，这是为什么呢？
很抱歉我上面的损失图应该是放错了，下面这个是大于10的情况，图中应该是没有两条曲线的，并且查看函数输出中的train.dat发现为空
“losshistory, train_state = model.train()
dde.saveplot(
    losshistory,
    train_state,
    issave=True,
    isplot=True,
    loss_fname="loss.dat",
    train_fname="train.dat",
    test_fname="test.dat",
)”

![QQ图片20220714163753](https://user-images.githubusercontent.com/73152456/178963776-b73af1a4-c30a-4f4c-aa19-66595297de74.png)

net.apply_feature_transform中当数值大于10时
2.通过硬边界和软边界的约束可以得到形状正确的输出，但是由于硬边界对输出进行转换，所以数据精度发生改变
“def transform(x, y):
res = x[:, 0:1] * (a - x[:, 0:1]) * x[:, 1:2] * (b - x[:, 1:2])
return res * y”
如果我想在res处加上一个系数保证输出的准确性，那么这个系数改如何确定比较好呢，是通过训练的输出自行调整还是有比较合适的确定方法？
或者除了添加系数以外，您有什么其他的建议吗？
下面是我的程序
`def ddw(x,y):
    t1=dde.grad.hessian(y,x,i=0,j=0)
    t2=dde.grad.hessian(y,x,i=1,j=1)
    return t1,t2

def pde(x,y):

    dw_xx = dde.grad.hessian(y, x, i=0, j=0)
    dw_yy = dde.grad.hessian(y, x, i=1, j=1)
    dw_xxxx = dde.grad.hessian(dw_xx, x, i=0, j=0)
    dw_xxyy = dde.grad.hessian(dw_xx, x, i=1, j=1)
    dw_yyyy = dde.grad.hessian(dw_yy, x, i=1, j=1)
    return dw_xxxx + 2 * dw_xxyy + dw_yyyy - q / D

“尝试加入硬边界约束”
def transform(x, y):
    res = x[:, 0:1] * (a - x[:, 0:1]) * x[:, 1:2] * (b - x[:, 1:2])
    return res * y


def boundary(x, on_boundary):
    return on_boundary

def boundary_x(x,on_boundary):
    return on_boundary and np.isclose(x[0],0 or a)

def boundary_y(x,on_boundary):
    return on_boundary and np.isclose(x[1],0 or b)

geom=dde.geometry.geometry_2d.Rectangle(xmin=[0,0], xmax=[a,b])


bc_1=dde.icbc.DirichletBC(geom,lambda x:0,boundary)
bc_x=dde.icbc.OperatorBC(geom,lambda x,y,_:ddw(x,y)[0],boundary_x)
bc_y=dde.icbc.OperatorBC(geom,lambda x,y,_:ddw(x,y)[1],boundary_y)


data = dde.data.PDE(geom, pde, [bc_1,bc_x,bc_y], num_domain=1200, num_boundary=240, num_test=1500)
net = dde.nn.FNN([2] + [50] * 4 + [1], "tanh", "Glorot uniform")

net.apply_feature_transform(lambda X: tf.concat([X[:,0:1]/a, X[:,1:2]/b], axis=1))

“硬边界输出转换”
net.apply_output_transform(transform)

model = dde.Model(data, net)

model.compile("adam", lr=0.001)
model.train(epochs=5000)
model.compile("L-BFGS-B")

losshistory, train_state = model.train()
dde.saveplot(
    losshistory,
    train_state,
    issave=True,
    isplot=True,
    loss_fname="loss.dat",
    train_fname="train.dat",
    test_fname="test.dat",
)`

**Reply by lululxvi** [reply]
1. The two lines are overlapped. 
2. I cannot understand your question "如果我想在res处加上一个系数保证输出的准确性，那么这个系数改如何确定比较好呢，是通过训练的输出自行调整还是有比较合适的确定方法？"
